qutip/tests/test_interpolate.py

```diff
@@ -30,325 +30,123 @@
+    was originally needed because the `brmesolve` solver cannot mix splines and
+    functions in the Hamiltonian (it can only accept string-formatted
+    time-dependence), but we want to parametrize this for all the other tests.
+    """
+    if ("solver" in metafunc.fixturenames
+            and "coefficients" in metafunc.fixturenames):
+        _parametrize_solver_coefficients(metafunc)
+
+
+@pytest.mark.parametrize('noise', [
+    pytest.param(lambda n: 0.1*np.random.rand(n), id="real"),
+    pytest.param(lambda n: 0.1j*np.random.rand(n), id="complex"),
+])
+def test_equivalence_to_scipy(noise):
+    x = np.linspace(0, 2*np.pi, 200)
+    y = np.sin(x) + noise(x.shape[0])
+    test = qutip.Cubic_Spline(x[0], x[-1], y)
+    expected = scipy.interpolate.CubicSpline(x, y, bc_type="natural")
+    # We use the bc_type="natural", i.e. zero second-derivatives at the
+    # boundaries because that matches qutip.
+    np.testing.assert_allclose(test(x), expected(x), atol=1e-10)
+    centres = 0.5 * (x[:-1] + x[1:])
+    # Test some points not in the original.
+    for point in centres:
+        assert np.abs(test(point) - expected(point)) < 1e-10, "Scalar argument"
+
+
+# We make a bunch of different test cases for the coefficients, all as a
+# three-tuple of
+#   (string, other, type)
+# where `string` is the string-format time-dependence (we use this as the
+# reference), `other` is what will become the time-dependence for this part of
+# the test, and `type` is a string in ["spline", "function", "string"] which
+# tells us how to interpret the type of `other`.  If `type` is "function" or
+# "string", then nothing is done to `other` and it is used as-is.  If `type` is
+# "spline", then `other` is given as a vectorised function which we turn into a
+# qutip.Cubic_Spline using appropriate `x` and `y` values.
+_case_real_spline = ('0.25*sin(t)', lambda t: 0.25*np.sin(t), 'spline')
+_case_real_string = ('0.25*cos(t)', '0.25*cos(t)', 'string')
+_case_real_function = ('0.25*sin(t)', lambda t, *_: 0.25*np.sin(t), 'function')
+_case_complex_spline = ('0.1j*sin(t)', lambda t: 0.1j*np.sin(t), 'spline')
+_case_complex_string = ('-0.1j*sin(t)', '-0.1j*sin(t)', 'string')
+
+
+def _parametrize_solver_coefficients(metafunc):
+    """
+    Perform the parametrisation for test cases using a solver and some
+    time-dependent coefficients.  This is necessary because not all solvers can
+    accept all combinations of time-dependence specifications.
+    """
+    solvers = [
+        (qutip.sesolve, 'sesolve'),
+        (qutip.mesolve, 'mesolve'),
+        (functools.partial(qutip.mcsolve, ntraj=500), "mcsolve"),
+        (qutip.brmesolve, 'brmesolve')
+    ]
+    coefficients = [
+        ([_case_real_spline], "real-spline"),
+        ([_case_real_spline, _case_real_string], "real-spline,string"),
+        ([_case_real_spline, _case_real_function], "real-spline,function"),
+        ([_case_real_function], "real-function"),
```

Comment:
`_case_real_function` does not use spline, does it need to be included here?

Reply:
No, it can go.

This may have been a mistake that it was committed - originally I was getting failures on these cases, and I had to debug it.  It's possible I forgot to remove that case if it wasn't in the previous batch.

Reply:
Fixed in #1250.